File: packages/shared/src/money.ts
/**
 * The single money abstraction for the platform (CLAUDE.md rule 1, ADR-7).
 *
 * Money is stored and compared as **integer paisa** (1 BDT = 100 paisa). It is
 * parsed from decimal strings and rendered back to 2-decimal strings on the
 * wire. No float arithmetic ever touches an amount.
 */

const PAISA_PER_TAKA = 100;
/** NUMERIC(14,2) → max 12 integer digits + 2 decimals → 99,999,999,999,999 paisa. */
const MAX_ABS_PAISA = 99_999_999_999_999;

const BENGALI_DIGIT_ZERO = 0x09e6; // '০'
const ASCII_DIGIT_ZERO = 0x30; // '0'

/** Map Bengali numerals (০–৯) to ASCII (0–9); leave everything else untouched. */
function toAsciiDigits(input: string): string {
  let out = '';
  for (const ch of input) {
    const code = ch.codePointAt(0);
    if (code !== undefined && code >= BENGALI_DIGIT_ZERO && code <= BENGALI_DIGIT_ZERO + 9) {
      out += String.fromCharCode(ASCII_DIGIT_ZERO + (code - BENGALI_DIGIT_ZERO));
    } else {
      out += ch;
    }
  }
  return out;
}

export class MoneyParseError extends Error {
  constructor(input: string, reason: string) {
    super(`Cannot parse money from ${JSON.stringify(input)}: ${reason}`);
    this.name = 'MoneyParseError';
  }
}

export class Money {
  private constructor(private readonly paisaValue: number) {}

  /** Construct from a known-integer paisa value. */
  static fromPaisa(paisa: number): Money {
    if (!Number.isInteger(paisa)) {
      throw new MoneyParseError(String(paisa), 'paisa must be an integer');
    }
    if (Math.abs(paisa) > MAX_ABS_PAISA) {
      throw new MoneyParseError(String(paisa), 'amount out of range for NUMERIC(14,2)');
    }
    return new Money(paisa);
  }

  /**
   * Parse a decimal string. Tolerates thousands separators, `Tk`/`BDT`/`৳`
   * prefixes and Bengali digits. Throws `MoneyParseError` on any ambiguity —
   * it must never silently return 0 or round (`1.005` is an error, not `1.00`).
   */
  static fromDecimalString(input: string): Money {
    const normalized = toAsciiDigits(input).trim();
    if (normalized.length === 0) {
      throw new MoneyParseError(input, 'empty');
    }
    const cleaned = normalized
      .replace(/^(tk\.?|bdt|৳)\s*/i, '')
      .replace(/\s*(tk|bdt|৳)$/i, '')
      .replace(/[\s,]/g, '');
    // Wire amounts are non-negative; a signed delta only ever comes from subtract().
    if (!/^\d+(\.\d{1,2})?$/.test(cleaned)) {
      throw new MoneyParseError(input, 'not a valid non-negative amount with up to 2 decimals');
    }
    const dot = cleaned.indexOf('.');
    const intPart = dot === -1 ? cleaned : cleaned.slice(0, dot);
    const fracRaw = dot === -1 ? '' : cleaned.slice(dot + 1);
    const fracPart = `${fracRaw}00`.slice(0, 2);
    return Money.fromPaisa(Number(intPart) * PAISA_PER_TAKA + Number(fracPart));
  }

  /** Accepts a Prisma Decimal, a number, or a string. */
  static fromPrismaDecimal(value: string | number | { toString(): string }): Money {
    return Money.fromDecimalString(String(value));
  }

  static zero(): Money {
    return new Money(0);
  }

  toPaisa(): number {
    return this.paisaValue;
  }

  /** Always two decimal places, e.g. `1250.00`. */
  toDecimalString(): string {
    const negative = this.paisaValue < 0;
    const abs = Math.abs(this.paisaValue);
    const int = Math.floor(abs / PAISA_PER_TAKA);
    const frac = abs % PAISA_PER_TAKA;
    return `${negative ? '-' : ''}${int}.${frac.toString().padStart(2, '0')}`;
  }

  equals(other: Money): boolean {
    return this.paisaValue === other.paisaValue;
  }

  compare(other: Money): -1 | 0 | 1 {
    if (this.paisaValue < other.paisaValue) return -1;
    if (this.paisaValue > other.paisaValue) return 1;
    return 0;
  }

  absDiff(other: Money): Money {
    return new Money(Math.abs(this.paisaValue - other.paisaValue));
  }

  /** True when |this − other| ≤ |tolerance|. */
  isWithinTolerance(other: Money, tolerance: Money): boolean {
    return this.absDiff(other).paisaValue <= Math.abs(tolerance.paisaValue);
  }

  add(other: Money): Money {
    return Money.fromPaisa(this.paisaValue + other.paisaValue);
  }

  subtract(other: Money): Money {
    return Money.fromPaisa(this.paisaValue - other.paisaValue);
  }

  isPositive(): boolean {
    return this.paisaValue > 0;
  }
}

File: apps/android/app/src/main/kotlin/com/inovisolutions/paymentsync/domain/model/Money.kt
package com.inovisolutions.paymentsync.domain.model

/**
 * Money as integer paisa (1 BDT = 100 paisa), mirroring `packages/shared/money.ts`
 * (CLAUDE.md rule 1). NEVER `Double` anywhere near an amount. Parses decimal
 * strings (tolerating thousands separators, `Tk`/`BDT`/`৳`, Bengali digits) and
 * renders back to a two-decimal string on the wire.
 */
@JvmInline
value class Money private constructor(val paisa: Long) {

    fun toDisplayString(): String {
        val negative = paisa < 0
        val abs = if (negative) -paisa else paisa
        val int = abs / 100
        val frac = abs % 100
        return buildString {
            if (negative) append('-')
            append(int)
            append('.')
            if (frac < 10) append('0')
            append(frac)
        }
    }

    companion object {
        private const val PAISA_PER_TAKA = 100L
        private const val MAX_ABS_PAISA = 99_999_999_999_999L

        fun fromPaisa(paisa: Long): Money {
            require(paisa in -MAX_ABS_PAISA..MAX_ABS_PAISA) { "amount out of range for NUMERIC(14,2)" }
            return Money(paisa)
        }

        fun zero(): Money = Money(0)

        /** Bengali numerals (০–৯, U+09E6..U+09EF) → ASCII; everything else untouched. */
        private fun toAsciiDigits(input: String): String = buildString {
            for (ch in input) {
                val code = ch.code
                if (code in 0x09E6..0x09EF) append('0' + (code - 0x09E6)) else append(ch)
            }
        }

        private val VALID = Regex("^\\d+(\\.\\d{1,2})?$")

        /** Throws on any ambiguity — `1.005` is an error, never silently rounded. */
        fun fromDecimalString(input: String): Money {
            val normalized = toAsciiDigits(input).trim()
            require(normalized.isNotEmpty()) { "empty amount" }
            val cleaned = normalized
                .replace(Regex("^(?i)(tk\\.?|bdt|৳)\\s*"), "")
                .replace(Regex("\\s*(?i)(tk|bdt|৳)$"), "")
                .replace(Regex("[\\s,]"), "")
            require(VALID.matches(cleaned)) { "not a valid non-negative amount: $input" }
            val dot = cleaned.indexOf('.')
            val intPart = if (dot == -1) cleaned else cleaned.substring(0, dot)
            val fracRaw = if (dot == -1) "" else cleaned.substring(dot + 1)
            val fracPart = (fracRaw + "00").substring(0, 2)
            return fromPaisa(intPart.toLong() * PAISA_PER_TAKA + fracPart.toLong())
        }

        /** Lenient parse used by the local parser hint — returns null instead of throwing. */
        fun fromDecimalStringOrNull(input: String): Money? =
            runCatching { fromDecimalString(input) }.getOrNull()
    }
}
